ClassFieldTheory/GroupCohomology/_9_CyclicGroup.lean

```diff
@@ -77,90 +99,206 @@ lemma map₁_comm (g : G) :
 lemma map₁_comp_coind_ι :
     map₁ (R := R) (G := G) (A := A) ∘ₗ coind₁'_ι = 0 := by
   ext; simp
 
-omit [Finite G] [DecidableEq G] in
 lemma map₁_ker :
-    LinearMap.ker (map₁ (R := R) (G := G) (A := A)) = LinearMap.range coind₁'_ι :=
-  sorry
+    LinearMap.ker (map₁ (R := R) (G := G) (A := A)) = LinearMap.range coind₁'_ι := by
+  ext f
+  constructor
+  · intro hf
+    rw [LinearMap.mem_ker] at hf
+    simp only [coind₁'_ι, LinearMap.mem_range, LinearMap.coe_mk, AddHom.coe_mk, ] at hf ⊢
+    use f (gen G)⁻¹
+    ext x
+    obtain ⟨n, hx⟩ : x ∈ Subgroup.zpowers (gen G) := IsCyclic.exists_generator.choose_spec x
+    dsimp at hx
+    rw [Function.const_apply, ← hx]
+    dsimp [map₁] at hf
+    induction n generalizing x with
+    | zero =>
+        rw [zpow_zero]
+        have := congr_fun hf 1
+        simp only [mul_one, Pi.zero_apply] at this
+        rw [sub_eq_zero] at this
+        exact this.symm
+    | succ n hn =>
+      have := congr_fun hf ((gen G ^ ((n : ℤ) + 1)))
+      simp only [Pi.zero_apply, sub_eq_zero] at this
+      rw [this, hn _ rfl]
+      group
+    | pred n hn =>
+      have := congr_fun hf ((gen G ^ (- (n : ℤ))))
+      simp only [Pi.zero_apply, sub_eq_zero] at this
+      rw [zpow_sub_one (gen G) _, hn _ rfl, this]
+      group
+  · rintro ⟨_, rfl⟩
+    exact LinearMap.congr_fun map₁_comp_coind_ι _
 
 @[simps!] def map₂ : (G →₀ A) →ₗ[R] (G →₀ A) :=
   LinearMap.id - lmapDomain _ _ (fun x ↦ gen G * x)
 
+omit [Finite G] [DecidableEq G]
+lemma map₂_apply (f : G →₀ A) (x : G) :
+    Representation.map₂ (R := R) f x = f x - f ((gen G)⁻¹ * x) := by
+  simp [Representation.map₂]
+  convert Finsupp.mapDomain_apply ?_ _ ((gen G)⁻¹ * x)
+  · simp
+  · intro x y h
+    simpa using h
+
 omit [Finite G] [DecidableEq G] in
 @[simp] lemma map₂_comp_lsingle (x : G) :
     map₂ (R := R) (G := G) (A := A) ∘ₗ lsingle x = lsingle x - lsingle (gen G * x) := by
   ext
   simp [map₂, LinearMap.sub_comp]
 
-omit [Finite G] [DecidableEq G] in
 lemma map₂_comm (g : G) :
     map₂ ∘ₗ ρ.ind₁' g = ρ.ind₁' g ∘ₗ map₂ := by
   ext x : 1
   rw [LinearMap.comp_assoc, ind₁'_comp_lsingle, LinearMap.comp_assoc, map₂_comp_lsingle,
     LinearMap.comp_sub, ind₁'_comp_lsingle, ←LinearMap.comp_assoc, map₂_comp_lsingle,
     LinearMap.sub_comp, ind₁'_comp_lsingle, mul_assoc]
 
-omit [Finite G] [DecidableEq G] in
 lemma ind₁'_π_comp_map₂ :
     ind₁'_π ∘ₗ map₂ (R := R) (G := G) (A := A) = 0 := by
   ext : 1
   rw [LinearMap.comp_assoc, map₂_comp_lsingle, LinearMap.comp_sub,
     LinearMap.zero_comp, sub_eq_zero, ind₁'_π_comp_lsingle, ind₁'_π_comp_lsingle]
 
-lemma map₂_range :
-    LinearMap.range (map₂ (R := R) (G := G) (A := A)) = LinearMap.ker ind₁'_π :=
-  sorry
+lemma map₂_range [Fintype G] [DecidableEq G] :
+    LinearMap.range (map₂ (R := R) (G := G) (A := A)) = LinearMap.ker ind₁'_π := by
+  ext w
+  constructor
+  · rintro ⟨y, rfl⟩
+    have := ind₁'_π_comp_map₂ (R := R) (G := G) (A := A)
+    apply_fun (· y) at this
+    exact this
```

Comment:
That can surely be golfed to one line? Maybe `exact congr_fun ind₁'_π_comp_map₂ y` or something?